luigi/scheduler.py

```diff
@@ -567,7 +567,8 @@ def prune(self):
 
         for task in self._state.get_active_tasks():
             self._state.fail_dead_worker_task(task, self._config, assistant_ids)
-            if task.id not in necessary_tasks and self._state.prune(task, self._config):
+            removed = self._state.prune(task, self._config)
+            if removed and task.id not in necessary_tasks:
```

Comment:
Ah, so the change is that you want the `prune()` side effect before doing the `task.id not in necessary_tasks` membership check?

Stupid non-Haskell programming languages :p